### PR TITLE
Fix bmm_sparse_cuda illegal memory access

### PR DESCRIPTION
**Pull request by ludvb**
This PR fixes a bug in `search_end_matrix_indices_cuda_kernel` causing an illegal memory access when calling `bmm_sparse_cuda` on a sparse matrix with no non-zero values in the first batch dimension. Reproducible example:
```py
import torch

ind = torch.tensor([[1], [0], [0]], device="cuda")
val = torch.tensor([1.], device="cuda")
A = torch.sparse_coo_tensor(ind, val, size=(2, 1, 1))
B = torch.zeros((2, 1, 1), device="cuda")
C = torch.bmm(A, B)
```

## Details

In the previous code, we may for example end up with the following situation:
```
i : indices_1D[i]
------------------------------------------
0 : 1                <- start_idx, mid_idx
1 : 1                <- end_idx
...
```
When `target_mat_num = 0`, the next iteration of the while loop will assign `-1` to `end_idx` and thus `(0 + (-1)) >> 1 = -1` to `mid_idx`, causing an access error on line 703. The updated code maintains the invariant `start_idx <= end_idx` and will not go out of bounds.

cc @alexsamardzic @nikitaved @pearu @cpuhrsch @amjames @bhosmer @jcaip

### CODE REVIEW

**Comment by Skylion007** [comment]
Can we add the test case in the example as a unit test?

**Comment by ludvb** [comment]
I've added a unit test now @Skylion007

File: test/test_sparse.py
Comment (Skylion007) [inline test/test_sparse.py:1458]: Do we not already have a more specific condition for this? Like CUSPARSE_SPMM_COMPLEX128_SUPPORTED

File: test/test_sparse.py
Comment (amjames) [inline test/test_sparse.py:1461]: It should be pretty simple to add this as a case for [`test_bmm` above](https://github.com/pytorch/pytorch/blob/ea42027e0ed7530386ae4222dc599fcaf84a8a05/test/test_sparse.py#L1367)

Since you need one sub matrix with no nonzero values, adding a test case with a single nnz and `num_mats > 1` should be able to trigger it?

**Review by amjames — COMMENTED** [review]
Couple of comments related to the test but overall looks like a nice simplification of the search logic here.

File: test/test_sparse.py
Comment (ludvb) [reply test/test_sparse.py:1469]: You're right, it would be good to have an explicit failure condition (the current version fails implicitly by raising a RuntimeError on the call to torch.bmm). Maybe we can do something like this?
```py
indices = torch.tensor([[1], [0], [0]], device=device)
values = torch.tensor([1.], device=device)
a = torch.sparse_coo_tensor(indices, values, size=(2, 1, 1))
b = torch.zeros((2, 1, 1), device=device)
try:
    _ = torch.bmm(a, b)
except RuntimeError as e:
    self.fail(f"torch.bmm failed: {e}")
```

File: test/test_sparse.py
Comment (ludvb) [reply test/test_sparse.py:1461]: Would also prefer not to have a separate test case. The thing that makes this a bit tricky is that the error is silent if the memory before the index tensor has been allocated since CUDA doesn't do any bounds checking (hence the call to `torch.cuda.empty_cache` at the beginning of the test). So we need to be extra careful in how we set things up in order for the error to trigger reliably (can't use `self._gen_sparse`). For example, just copy-pasting the current test to the end of `test_bmm` results in the exception no longer being raised. This can be fixed by either having the test at the top of `test_bmm` or keeping it at the end but prepending it with `del a, b, ab, ab_mat, ab_mat_check, ab_traspose_check`. However, I feel both solutions could easily break in the future, so I'm not sure what the best option would be?

**Comment by ludvb** [comment]
Appreciate the feedback @amjames, I've responded inline with a few follow-up comments and questions.

There's another problem with how the test works that I didn't realize before: The IMA makes CUDA enter an error state that causes subsequent tests on the same CUDA device to also fail. It's currently not possible to recover from this state as far as I can tell (related discussion in #72117), so I'm not sure if triggering an IMA is the right way to go. Unfortunately, I've not found another way to test for this error - as long as the old algorithm stays within bounds, it does return the correct result.

File: test/test_sparse.py
Comment (amjames) [reply test/test_sparse.py:1469]: It would be enough to assert on the result being correct (easy to construct the expected result by hand here). With a comment that refers to this issue for reference.

File: test/test_sparse.py
Comment (amjames) [reply test/test_sparse.py:1461]: Ah, I didn't realize that it was so sensitive. In that case a stand-alone test like this is warranted.

**Comment by amjames** [comment]
> There's another problem with how the test works that I didn't realize before: The IMA makes CUDA enter an error state that causes subsequent tests on the same CUDA device to also fail. It's currently not possible to recover from this state as far as I can tell (related discussion in #72117), so I'm not sure if triggering an IMA is the right way to go. Unfortunately, I've not found another way to test for this error - as long as the old algorithm stays within bounds, it does return the correct result.

This PR is a fix for the IMA, so the test would only trigger that if a future modification regressed in some way. I would say that is the intended purpose of the test. The follow on failures are unavoidable, but it still does the job of signaling if some other change reintroduces the bug. I don't see any problem with that.

**Comment by ludvb** [comment]
Got it, that makes sense @amjames. I've added an assert statement to the test now.

**Review by amjames — APPROVED** [review]
I'm happy with this now. We will still need a review from someone on the "approved" list before merging though.

**Review by pearu — APPROVED** [review]
I confirm that the new search_end_matrix_indices algorithm is correct. Thanks, @ludvb!

@ludvb could rebase the PR against main branch?

**Comment by ludvb** [comment]
Thanks for the reviews! I've rebased the code @pearu

File: test/test_sparse.py
Comment (nikitaved) [reply test/test_sparse.py:1461]: What about a comment explaining the need  for a standalone test? Seems rather non-trivial

**Review by nikitaved — APPROVED** [review]

File: test/test_sparse.py
Comment (ludvb) [reply test/test_sparse.py:1461]: Thanks for the review @nikitaved! I've added a comment to the test

**Comment by ludvb** [comment]
@Skylion007 @amjames Do we need to rerun workflows? Sorry for ping - not quite sure how this works

**Comment by ludvb** [comment]
@pearu can you help me with approving the workflows again? :pray: I've rebased on main, hopefully that will resolve the previous failures

**Pull request closed without merging**
